File: bindings/Octave/mlibkriging/tests/WarpKriging_test.m
%% Tests for WarpKriging Octave binding.
%% Mirrors bindings/Python/pylibkriging/tests/WarpKriging_test.py
%% and bindings/R/rlibkriging/tests/testthat/test-WarpKriging.R.
1;  % mark this file as a script

f1d = @(x) 1 - 0.5 * (sin(12 * x) ./ (1 + x) + 2 * cos(7 * x) .* x.^5 + 0.7);
branin = @(x1, x2) (x2 - 5.1 / (4 * pi^2) .* x1.^2 + 5.0 / pi .* x1 - 6.0).^2 ...
                   + 10 * (1 - 1.0 / (8 * pi)) .* cos(x1) + 10;

n_failed = 0;

% -----------------------------------------------------------------------
% Test 1: Kumaraswamy warping (1D continuous)
% -----------------------------------------------------------------------
try
    X = linspace(0.01, 0.99, 8)';
    y = f1d(X);

    k = WarpKriging(y, X, {"kumaraswamy"}, "gauss");
    s = k.summary();
    assert(ischar(s) && length(s) > 0);

    [mean_p, stdev_p] = k.predict(X, true, false, false);
    assert(length(mean_p) == 8);
    assert(all(isfinite(mean_p)));

    x_pred = linspace(0.01, 0.99, 50)';
    [mean2, ~] = k.predict(x_pred, true, false, false);
    rmse = sqrt(mean((mean2 - f1d(x_pred)).^2));
    assert(isfinite(rmse));
    fprintf("  Test 1 Kumaraswamy OK (RMSE=%.4f)\n", rmse);
catch err
    fprintf("  Test 1 FAILED: %s\n", err.message);
    n_failed = n_failed + 1;
end

% -----------------------------------------------------------------------
% Test 2: Categorical embedding
% -----------------------------------------------------------------------
try
    mu = [1.0; 5.0; 3.0];
    rand("state", 42);
    n = 15;
    levels = mod(0:(n-1), 3)';
    X = levels;
    y = mu(levels + 1) + 0.1 * randn(n, 1);

    k = WarpKriging(y, X, {"categorical(3,2)"}, "gauss");
    X_test = (0:2)';
    [mean_p, ~] = k.predict(X_test, true, false, false);
    assert(length(mean_p) == 3);
    assert(all(isfinite(mean_p)));
    fprintf("  Test 2 Categorical OK\n");
catch err
    fprintf("  Test 2 FAILED: %s\n", err.message);
    n_failed = n_failed + 1;
end

% -----------------------------------------------------------------------
% Test 3: Mixed continuous + categorical
% -----------------------------------------------------------------------
try
    offset = [1.0; 2.0; 0.5];
    rand("state", 99);
    n = 30;
    x1 = rand(n, 1);
    x2 = mod(0:(n-1), 3)';
    X = [x1, x2];
    y = sin(2 * pi * x1) .* offset(x2 + 1);

    k = WarpKriging(y, X, {"kumaraswamy", "categorical(3,2)"}, "matern5_2");
    xc = linspace(0.01, 0.99, 20)';
    for cat_idx = 0:2
        X_test = [xc, cat_idx * ones(20, 1)];
        [mean_p, ~] = k.predict(X_test, true, false, false);
        assert(length(mean_p) == 20);
        assert(all(isfinite(mean_p)));
    end
    fprintf("  Test 3 Mixed OK\n");
catch err
    fprintf("  Test 3 FAILED: %s\n", err.message);
    n_failed = n_failed + 1;
end

% -----------------------------------------------------------------------
% Test 4: Ordinal warping
% -----------------------------------------------------------------------
try
    L = 5;
    rand("state", 7);
    n = 20;
    levels = mod(0:(n-1), L)';
    X = levels;
    y = levels.^2 + 0.1 * randn(n, 1);

    k = WarpKriging(y, X, {"ordinal(5)"}, "gauss");
    X_test = (0:(L-1))';
    [mean_p, ~] = k.predict(X_test, true, false, false);
    assert(length(mean_p) == L);
    assert(all(isfinite(mean_p)));
    fprintf("  Test 4 Ordinal OK\n");
catch err
    fprintf("  Test 4 FAILED: %s\n", err.message);
    n_failed = n_failed + 1;
end

% -----------------------------------------------------------------------
% Test 5: NeuralMono
% -----------------------------------------------------------------------
try
    X = linspace(0.01, 0.99, 10)';
    y = f1d(X);
    k = WarpKriging(y, X, {"neural_mono(8)"}, "gauss");
    [mean_p, ~] = k.predict(X, true, false, false);
    assert(all(isfinite(mean_p)));
    fprintf("  Test 5 NeuralMono OK\n");
catch err
    fprintf("  Test 5 FAILED: %s\n", err.message);
    n_failed = n_failed + 1;
end

% -----------------------------------------------------------------------
% Test 6: MLP warping 1D
% -----------------------------------------------------------------------
try
    X = linspace(0.01, 0.99, 10)';
    y = f1d(X);
    k = WarpKriging(y, X, {"mlp(16:8,3,selu)"}, "gauss");
    x_pred = linspace(0.01, 0.99, 50)';
    [mean_p, ~] = k.predict(x_pred, true, false, false);
    rmse = sqrt(mean((mean_p - f1d(x_pred)).^2));
    assert(isfinite(rmse));
    fprintf("  Test 6 MLP 1D OK (RMSE=%.4f)\n", rmse);
catch err
    fprintf("  Test 6 FAILED: %s\n", err.message);
    n_failed = n_failed + 1;
end

% -----------------------------------------------------------------------
% Test 7: MLP + categorical mixed
% -----------------------------------------------------------------------
try
    offset = [1.0; 2.0; 0.5];
    rand("state", 99);
    n = 30;
    x1 = rand(n, 1);
    x2 = mod(0:(n-1), 3)';
    X = [x1, x2];
    y = sin(2 * pi * x1) .* offset(x2 + 1);

    k = WarpKriging(y, X, {"mlp(16:8,2,tanh)", "categorical(3,2)"}, "matern5_2");
    xc = linspace(0.01, 0.99, 20)';
    for cat_idx = 0:2
        X_test = [xc, cat_idx * ones(20, 1)];
        [mean_p, ~] = k.predict(X_test, true, false, false);
        assert(all(isfinite(mean_p)));
    end
    fprintf("  Test 7 MLP+Cat OK\n");
catch err
    fprintf("  Test 7 FAILED: %s\n", err.message);
    n_failed = n_failed + 1;
end

% -----------------------------------------------------------------------
% Test 8: simulate mixed
% -----------------------------------------------------------------------
try
    rand("state", 42);
    n = 20;
    x1 = rand(n, 1);
    x2 = mod(0:(n-1), 2)';
    X = [x1, x2];
    mult = [1.0; 3.0];
    y = sin(2 * pi * x1) .* mult(x2 + 1);

    k = WarpKriging(y, X, {"affine", "categorical(2,2)"}, "gauss");
    X_sim = [linspace(0.1, 0.9, 10)', zeros(10, 1)];
    sims = k.simulate(int32(30), int32(123), X_sim);
    assert(size(sims, 1) == 10);
    assert(size(sims, 2) == 30);
    assert(all(isfinite(sims(:))));
    fprintf("  Test 8 Simulate mixed OK\n");
catch err
    fprintf("  Test 8 FAILED: %s\n", err.message);
    n_failed = n_failed + 1;
end

% -----------------------------------------------------------------------
% Test 9: update
% -----------------------------------------------------------------------
try
    X0 = [0.1, 0.0; 0.5, 1.0; 0.9, 0.0];
    y0 = [1.0; 3.0; 0.5];
    k = WarpKriging(y0, X0, {"none", "categorical(2,1)"}, "gauss");
    X_new = [0.3, 1.0; 0.7, 0.0];
    k.update([2.0; 1.5], X_new);
    [mean_p, ~] = k.predict(X0, true, false, false);
    assert(all(isfinite(mean_p)));
    fprintf("  Test 9 Update OK\n");
catch err
    fprintf("  Test 9 FAILED: %s\n", err.message);
    n_failed = n_failed + 1;
end

% -----------------------------------------------------------------------
% Test 10: Branin 2D with per-variable MLP
% -----------------------------------------------------------------------
try
    rand("state", 77);
    n = 25;
    X = rand(n, 2);
    y = branin(X(:, 1) * 15 - 5, X(:, 2) * 15);

    k = WarpKriging(y, X, {"mlp(16:8,2,selu)", "mlp(16:8,2,selu)"}, ...
                    "matern5_2", "constant", true);
    rand("state", 88);
    X_test = rand(15, 2);
    [mean_p, stdev_p] = k.predict(X_test, true, false, false);
    assert(length(mean_p) == 15);
    assert(all(isfinite(stdev_p)));

    sims = k.simulate(int32(20), int32(42), X_test);
    assert(size(sims, 1) == 15);
    assert(size(sims, 2) == 20);
    fprintf("  Test 10 Branin MLP OK\n");
catch err
    fprintf("  Test 10 FAILED: %s\n", err.message);
    n_failed = n_failed + 1;
end

% -----------------------------------------------------------------------
% Test 11: None vs Kumaraswamy vs MLP comparison
% -----------------------------------------------------------------------
try
    X = linspace(0.01, 0.99, 12)';
    y = f1d(X);
    xp = linspace(0.01, 0.99, 50)';
    ytrue = f1d(xp);

    k_none = WarpKriging(y, X, {"none"}, "gauss");
    k_kuma = WarpKriging(y, X, {"kumaraswamy"}, "gauss");
    k_mlp  = WarpKriging(y, X, {"mlp(16:8,2,selu)"}, "gauss");

    [m_none, ~] = k_none.predict(xp, true, false, false);
    [m_kuma, ~] = k_kuma.predict(xp, true, false, false);
    [m_mlp, ~]  = k_mlp.predict(xp, true, false, false);
    for mm = {m_none, m_kuma, m_mlp}
        rmse = sqrt(mean((mm{1} - ytrue).^2));
        assert(isfinite(rmse));
    end
    assert(isfinite(k_none.logLikelihood()));
    assert(isfinite(k_kuma.logLikelihood()));
    assert(isfinite(k_mlp.logLikelihood()));
    fprintf("  Test 11 Comparison OK\n");
catch err
    fprintf("  Test 11 FAILED: %s\n", err.message);
    n_failed = n_failed + 1;
end

% -----------------------------------------------------------------------
% Test 12: logLikelihoodFun + FD gradient
% -----------------------------------------------------------------------
try
    X = linspace(0.01, 0.99, 8)';
    y = f1d(X);
    k = WarpKriging(y, X, {"affine"}, "gauss", "constant", false, "none");

    % Evaluate at a fixed benign theta (not the optimum) so FD is well-conditioned.
    th = 0.3 * ones(size(k.theta()));
    [ll, grad] = k.logLikelihoodFun(th, true, false);
    assert(isfinite(ll));
    assert(all(isfinite(grad)));

    h = 1e-5;
    grad_num = zeros(size(th));
    for i = 1:length(th)
        tp = th; tm = th;
        tp(i) = tp(i) + h;
        tm(i) = tm(i) - h;
        [lp, ~] = k.logLikelihoodFun(tp, false, false);
        [lm, ~] = k.logLikelihoodFun(tm, false, false);
        grad_num(i) = (lp - lm) / (2 * h);
    end
    rel = norm(grad(:) - grad_num(:)) / (norm(grad_num(:)) + 1e-12);
    assert(rel < 1e-4, sprintf("grad FD rel error %.3g", rel));
    fprintf("  Test 12 logLikelihoodFun FD OK (rel=%.3g)\n", rel);
catch err
    fprintf("  Test 12 FAILED: %s\n", err.message);
    n_failed = n_failed + 1;
end

% -----------------------------------------------------------------------
% Test 13: Accessors / summary
% -----------------------------------------------------------------------
try
    X = linspace(0.01, 0.99, 6)';
    y = f1d(X);
    k = WarpKriging(y, X, {"boxcox"}, "matern3_2");

    s = k.summary();
    assert(ischar(s) && length(s) > 0);
    th = k.theta();
    assert(length(th) > 0 && all(th > 0));
    assert(k.sigma2() > 0);
    assert(strcmp(k.kernel(), "matern3_2"));
    assert(isfinite(k.logLikelihood()));
    ws = k.warping();
    assert(iscell(ws) && length(ws) == 1 && strcmp(ws{1}, "boxcox"));
    assert(k.is_fitted());
    fprintf("  Test 13 Accessors OK\n");
catch err
    fprintf("  Test 13 FAILED: %s\n", err.message);
    n_failed = n_failed + 1;
end

% -----------------------------------------------------------------------
% Test 14: predict with derivatives
% -----------------------------------------------------------------------
try
    X = linspace(0.01, 0.99, 8)';
    y = f1d(X);
    k = WarpKriging(y, X, {"affine"}, "gauss");
    X_new = linspace(0.1, 0.9, 5)';
    [mean_p, stdev_p, cov_p, mean_deriv, stdev_deriv] = k.predict(X_new, true, false, true);
    assert(size(mean_deriv, 1) == 5);
    assert(size(mean_deriv, 2) == 1);
    assert(size(stdev_deriv, 1) == 5);
    assert(size(stdev_deriv, 2) == 1);
    assert(all(isfinite(mean_deriv(:))));
    assert(all(isfinite(stdev_deriv(:))));
    fprintf("  Test 14 Predict-deriv OK\n");
catch err
    fprintf("  Test 14 FAILED: %s\n", err.message);
    n_failed = n_failed + 1;
end

% -----------------------------------------------------------------------
% Test 15: Getters (X, y, feature_dim)
% -----------------------------------------------------------------------
try
    X = linspace(0.01, 0.99, 8)';
    y = f1d(X);
    k = WarpKriging(y, X, {"affine"}, "gauss");
    Xr = k.X();
    assert(size(Xr, 1) == 8);
    assert(size(Xr, 2) == 1);
    yr = k.y();
    assert(length(yr) == 8);
    assert(k.feature_dim() > 0);
    fprintf("  Test 15 Getters OK\n");
catch err
    fprintf("  Test 15 FAILED: %s\n", err.message);
    n_failed = n_failed + 1;
end

% -----------------------------------------------------------------------
% Test 16: Copy
% -----------------------------------------------------------------------
try
    X = linspace(0.01, 0.99, 8)';
    y = f1d(X);
    k = WarpKriging(y, X, {"affine"}, "gauss");
    k2 = k.copy();
    assert(k2.is_fitted());
    assert(strcmp(k2.kernel(), k.kernel()));
    X_test = linspace(0.1, 0.9, 5)';
    [m1, ~] = k.predict(X_test, true, false, false);
    [m2, ~] = k2.predict(X_test, true, false, false);
    assert(max(abs(m1 - m2)) < 1e-10);
    fprintf("  Test 16 Copy OK\n");
catch err
    fprintf("  Test 16 FAILED: %s\n", err.message);
    n_failed = n_failed + 1;
end

if n_failed > 0
    error("WarpKriging tests: %d failed", n_failed);
end
fprintf("WarpKriging tests: all 16 passed\n");

File: bindings/Julia/jlibkriging/tests/warp_kriging_test.jl
using Test
using jlibkriging

f_test(x) = 1.0 - 0.5 * (sin(12.0 * x) / (1.0 + x) + 2.0 * cos(7.0 * x) * x^5 + 0.7)

function branin(x1, x2)
    a, b, cc = 1.0, 5.1 / (4 * pi^2), 5.0 / pi
    r, s, t = 6.0, 10.0, 1.0 / (8 * pi)
    return a * (x2 - b * x1^2 + cc * x1 - r)^2 + s * (1 - t) * cos(x1) + s
end

@testset "WarpKriging" begin

    @testset "Kumaraswamy 1D" begin
        X = reshape(collect(range(0.01, 0.99; length=8)), :, 1)
        y = [f_test(x) for x in X[:, 1]]

        k = WarpKriging(y, X, ["kumaraswamy"], "gauss";
                        parameters=Dict("max_iter_adam" => "200"))
        s = jlibkriging.summary(k)
        @test isa(s, String) && length(s) > 0

        result = predict(k, X)
        @test length(result.mean) == 8
        @test all(isfinite.(result.mean))

        x_pred = reshape(collect(range(0.01, 0.99; length=50)), :, 1)
        result2 = predict(k, x_pred)
        y_true = [f_test(x) for x in x_pred[:, 1]]
        rmse = sqrt(sum((result2.mean .- y_true).^2) / length(y_true))
        @test isfinite(rmse)
    end

    @testset "Categorical embedding" begin
        mu = [1.0, 5.0, 3.0]
        n = 15
        levels = repeat([0, 1, 2], ceil(Int, n / 3))[1:n]
        X = reshape(Float64.(levels), :, 1)
        y = [mu[levels[i]+1] + 0.1 * sin(Float64(i)) for i in 1:n]

        k = WarpKriging(y, X, ["categorical(3,2)"], "gauss";
                        parameters=Dict("max_iter_adam" => "200"))

        X_test = reshape([0.0, 1.0, 2.0], :, 1)
        result = predict(k, X_test)
        @test length(result.mean) == 3
        @test all(isfinite.(result.mean))
    end

    @testset "Mixed continuous + categorical" begin
        offset = [1.0, 2.0, 0.5]
        n = 30
        x1 = collect(range(0.05, 0.95; length=n))
        x2 = repeat([0, 1, 2], ceil(Int, n / 3))[1:n]
        X = hcat(x1, Float64.(x2))
        y = [sin(2 * pi * x1[i]) * offset[x2[i]+1] for i in 1:n]

        k = WarpKriging(y, X, ["kumaraswamy", "categorical(3,2)"], "matern5_2";
                        parameters=Dict("max_iter_adam" => "300"))

        xc = collect(range(0.01, 0.99; length=20))
        for cat_idx in 0:2
            X_test = hcat(xc, fill(Float64(cat_idx), 20))
            result = predict(k, X_test)
            @test length(result.mean) == 20
            @test all(isfinite.(result.mean))
        end
    end

    @testset "Ordinal" begin
        L = 5
        n = 20
        levels = repeat(collect(0:L-1), ceil(Int, n / L))[1:n]
        X = reshape(Float64.(levels), :, 1)
        y = Float64.(levels .^ 2) .+ [0.1 * sin(Float64(i)) for i in 1:n]

        k = WarpKriging(y, X, ["ordinal(5)"], "gauss";
                        parameters=Dict("max_iter_adam" => "200"))

        X_test = reshape(Float64.(collect(0:L-1)), :, 1)
        result = predict(k, X_test)
        @test length(result.mean) == L
        @test all(isfinite.(result.mean))
    end

    @testset "Neural mono" begin
        X = reshape(collect(range(0.01, 0.99; length=10)), :, 1)
        y = [f_test(x) for x in X[:, 1]]

        k = WarpKriging(y, X, ["neural_mono(8)"], "gauss";
                        parameters=Dict("max_iter_adam" => "200"))

        result = predict(k, X)
        @test all(isfinite.(result.mean))
    end

    @testset "MLP warping 1D" begin
        X = reshape(collect(range(0.01, 0.99; length=10)), :, 1)
        y = [f_test(x) for x in X[:, 1]]

        k = WarpKriging(y, X, ["mlp(16:8,3,selu)"], "gauss";
                        parameters=Dict("max_iter_adam" => "300"))

        x_pred = reshape(collect(range(0.01, 0.99; length=50)), :, 1)
        result = predict(k, x_pred)
        y_true = [f_test(x) for x in x_pred[:, 1]]
        rmse = sqrt(sum((result.mean .- y_true).^2) / length(y_true))
        @test isfinite(rmse)
    end

    @testset "MLP + categorical mixed" begin
        offset = [1.0, 2.0, 0.5]
        n = 30
        x1 = collect(range(0.05, 0.95; length=n))
        x2 = repeat([0, 1, 2], ceil(Int, n / 3))[1:n]
        X = hcat(x1, Float64.(x2))
        y = [sin(2 * pi * x1[i]) * offset[x2[i]+1] for i in 1:n]

        k = WarpKriging(y, X, ["mlp(16:8,2,tanh)", "categorical(3,2)"], "matern5_2";
                        parameters=Dict("max_iter_adam" => "300"))

        xc = collect(range(0.01, 0.99; length=20))
        for cat_idx in 0:2
            X_test = hcat(xc, fill(Float64(cat_idx), 20))
            result = predict(k, X_test)
            @test length(result.mean) == 20
            @test all(isfinite.(result.mean))
        end
    end

    @testset "Branin 2D per-variable MLP" begin
        n = 25
        x1 = collect(range(0.05, 0.95; length=n))
        x2 = collect(range(0.05, 0.95; length=n))
        X = hcat(x1, x2)
        y = [branin(X[i, 1] * 15 - 5, X[i, 2] * 15) for i in 1:n]

        k = WarpKriging(y, X,
                        ["mlp(16:8,2,selu)", "mlp(16:8,2,selu)"], "matern5_2";
                        normalize=true,
                        parameters=Dict("max_iter_adam" => "300"))

        X_test = hcat(collect(range(0.1, 0.9; length=15)),
                      collect(range(0.1, 0.9; length=15)))
        result = predict(k, X_test)
        @test length(result.mean) == 15
        @test all(isfinite.(result.stdev))

        sims = simulate(k, 20, 42, X_test)
        @test size(sims) == (15, 20)
    end

    @testset "None vs kumaraswamy vs MLP comparison" begin
        X = reshape(collect(range(0.01, 0.99; length=12)), :, 1)
        y = [f_test(x) for x in X[:, 1]]
        xp = reshape(collect(range(0.01, 0.99; length=50)), :, 1)
        ytrue = [f_test(x) for x in xp[:, 1]]

        k_none = WarpKriging(y, X, ["none"], "gauss";
                             parameters=Dict("max_iter_adam" => "200"))
        k_kuma = WarpKriging(y, X, ["kumaraswamy"], "gauss";
                             parameters=Dict("max_iter_adam" => "200"))
        k_mlp = WarpKriging(y, X, ["mlp(16:8,2,selu)"], "gauss";
                            parameters=Dict("max_iter_adam" => "300"))

        for k in (k_none, k_kuma, k_mlp)
            result = predict(k, xp)
            rmse = sqrt(sum((result.mean .- ytrue).^2) / length(ytrue))
            @test isfinite(rmse)
            @test isfinite(log_likelihood(k))
        end
    end

    @testset "Simulate" begin
        n = 20
        x1 = collect(range(0.05, 0.95; length=n))
        x2 = repeat([0, 1], ceil(Int, n / 2))[1:n]
        X = hcat(x1, Float64.(x2))
        y = [sin(2 * pi * x1[i]) * [1.0, 3.0][x2[i]+1] for i in 1:n]

        k = WarpKriging(y, X, ["affine", "categorical(2,2)"], "gauss";
                        parameters=Dict("max_iter_adam" => "200"))

        X_sim = hcat(collect(range(0.1, 0.9; length=10)), zeros(10))
        sims = simulate(k, 30, 123, X_sim)
        @test size(sims) == (10, 30)
        @test all(isfinite.(sims))
    end

    @testset "Update" begin
        X0 = [0.1 0.0; 0.5 1.0; 0.9 0.0]
        y0 = [1.0, 3.0, 0.5]

        k = WarpKriging(y0, X0, ["none", "categorical(2,1)"], "gauss";
                        parameters=Dict("max_iter_adam" => "100"))

        X_new = [0.3 1.0; 0.7 0.0]
        update!(k, [2.0, 1.5], X_new)

        result = predict(k, X0)
        @test all(isfinite.(result.mean))
    end

    @testset "Predict with derivatives" begin
        X = reshape(collect(range(0.01, 0.99; length=10)), :, 1)
        y = [f_test(x) for x in X[:, 1]]

        k = WarpKriging(y, X, ["affine"], "gauss";
                        parameters=Dict("max_iter_adam" => "100"))

        X_new = reshape([0.1, 0.3, 0.5, 0.7, 0.9], :, 1)
        result = predict(k, X_new; return_deriv=true)
        @test size(result.mean_deriv) == (5, 1)
        @test size(result.stdev_deriv) == (5, 1)
        @test all(isfinite.(result.mean_deriv))
        @test all(isfinite.(result.stdev_deriv))
    end

    @testset "Predict with covariance" begin
        X = reshape(collect(range(0.01, 0.99; length=8)), :, 1)
        y = [f_test(x) for x in X[:, 1]]

        k = WarpKriging(y, X, ["affine"], "gauss";
                        parameters=Dict("max_iter_adam" => "100"))

        X_new = reshape([0.1, 0.3, 0.5, 0.7, 0.9], :, 1)
        result = predict(k, X_new; return_stdev=true, return_cov=true)
        @test result.cov !== nothing
        @test size(result.cov) == (5, 5)
        @test all(isfinite.(result.cov))
        @test maximum(abs.(result.cov - result.cov')) < 1e-10
    end

    @testset "Accessors and summary" begin
        X = reshape(collect(range(0.01, 0.99; length=6)), :, 1)
        y = [f_test(x) for x in X[:, 1]]

        k = WarpKriging(y, X, ["boxcox"], "matern3_2";
                        parameters=Dict("max_iter_adam" => "100"))

        s = jlibkriging.summary(k)
        @test isa(s, String) && length(s) > 0

        th = get_theta(k)
        @test length(th) > 0
        @test all(th .> 0)

        @test get_sigma2(k) > 0
        @test kernel(k) == "matern3_2"
        @test isfinite(log_likelihood(k))
        @test get_warping(k) == ["boxcox"]
        @test is_fitted(k)
    end

    @testset "Log-likelihood gradient" begin
        X = reshape(collect(range(0.01, 0.99; length=8)), :, 1)
        y = [f_test(x) for x in X[:, 1]]

        k = WarpKriging(y, X, ["affine"], "gauss"; optim="none")

        # Evaluate at a fixed benign theta (not the optimum) so FD is well-conditioned.
        th = fill(0.3, length(get_theta(k)))
        ll_res = log_likelihood_fun(k, th; return_grad=true)
        @test isfinite(ll_res.ll)
        @test all(isfinite.(ll_res.grad))

        # Numerical gradient check
        h = 1e-5
        grad_num = zeros(length(th))
        for i in 1:length(th)
            tp = copy(th)
            tm = copy(th)
            tp[i] += h
            tm[i] -= h
            ll_p = log_likelihood_fun(k, tp)
            ll_m = log_likelihood_fun(k, tm)
            grad_num[i] = (ll_p.ll - ll_m.ll) / (2 * h)
        end

        rel = sqrt(sum((ll_res.grad .- grad_num).^2)) / (sqrt(sum(grad_num.^2)) + 1e-12)
        @test rel < 1e-4
    end

    @testset "Empty constructor then fit" begin
        wk = WarpKriging(["affine"], "gauss")
        @test !is_fitted(wk)

        X = reshape(collect(range(0.01, 0.99; length=8)), :, 1)
        y = [f_test(x) for x in X[:, 1]]
        fit!(wk, y, X)
        @test is_fitted(wk)

        result = predict(wk, X)
        @test length(result.mean) == 8
        @test all(isfinite.(result.mean))
    end

    @testset "Copy" begin
        X = reshape(collect(range(0.01, 0.99; length=8)), :, 1)
        y = [f_test(x) for x in X[:, 1]]

        k = WarpKriging(y, X, ["kumaraswamy"], "gauss";
                        parameters=Dict("max_iter_adam" => "100"))

        k2 = Base.copy(k)
        @test is_fitted(k2)
        @test kernel(k2) == kernel(k)
        @test get_warping(k2) == get_warping(k)

        result1 = predict(k, X)
        result2 = predict(k2, X)
        @test all(isfinite.(result2.mean))
    end

    @testset "Getters" begin
        X = reshape(collect(range(0.01, 0.99; length=8)), :, 1)
        y = [f_test(x) for x in X[:, 1]]

        k = WarpKriging(y, X, ["kumaraswamy"], "gauss";
                        parameters=Dict("max_iter_adam" => "100"))

        @test size(get_X(k)) == (8, 1)
        @test length(get_y(k)) == 8
        @test length(get_theta(k)) > 0
        @test get_sigma2(k) > 0.0
        @test feature_dim(k) > 0
    end
end
